storage: implement disk-based sideloaded storage

tbg opened the pull request:
Tests are TODO. Out for early review since it's indispensable for serious
testing of RESTORE (the in-memory storage does not get ejected eagerly
enough since the Raft log queue is unaware of sideloaded data, so you
quickly get OOMs).

cc @danhhz

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16924)
<!-- Reviewable:end -->

tschottdorf commented on pkg/storage/replica.go:
```diff
@@ -618,6 +618,10 @@ func (r *Replica) init(
 ) error {
+	// Replicas without a replicaID don't have a sideloaded storage. If the
+	// replicaID is nontrivial, we'll recreate the storage later.
+	r.raftMu.sideloaded = nil
```
This actually breaks all sorts of things. Have to check what's really going on here, but it seems replicas process Raft commands (log truncation) before they have their sideloaded storage initialized.

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 695 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KoSmlK3P3UyGSnzqN8f:-KoSmlK3P3UyGSnzqN8g:boimny8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica.go#L695)):*
> ```Go
> 			return errors.Wrap(err, "while initializing sideloaded storage")
> 		}
> 	}
> ```

Does this belong in `setReplicaIDLocked`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM aside from the lack of tests. 

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 622 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohHPOOxabP3dZi-853:-KohHPOOxabP3dZi-854:bjuqxuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica.go#L622)):*
> ```Go
> ) error {
> 	// Replicas without a replicaID don't have a sideloaded storage. If the
> 	// replicaID is nontrivial, we'll recreate the storage later.
> ```

Nit: I'm not sure where this usage of "trivial" to mean "zero" came from, but it annoys me. 

---

*[pkg/storage/replica.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KoRMfajgq4-YZ7nCIve-r1-623:-KohITpkjaHt3EXg8Syi:b-r1pxi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica.go#L623)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This actually breaks all sorts of things. Have to check what's really going on here, but it seems replicas process Raft commands (log truncation) before they have their sideloaded storage initialized.
</blockquote></details>

How did we get a non-nil `r.raftMu.sideloaded` at the beginning of `init`?

---

*[pkg/storage/replica.go, line 695 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KoSmlK3P3UyGSnzqN8f:-KohQ6BaEhxXcNrXSTC8:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica.go#L695)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does this belong in `setReplicaIDLocked`?
</blockquote></details>

Yes.

---

*[pkg/storage/replica_sideload_disk.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohJdHaRalL7gDtTkSj:-KohJdHaRalL7gDtTkSk:bxqtvqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_disk.go#L41)):*
> ```Go
> ) (sideloadStorage, error) {
> 	dir := filepath.Join(baseDir, fmt.Sprintf("%d.%d", rangeID, replicaID))
> 	if err := os.MkdirAll(dir, 0700); err != nil {
> ```

For consistency with rocksdb, we should use permissions 0755 for directories and [0644 for files](https://github.com/facebook/rocksdb/blob/56656e12d67d8a63f1e4c4214da9feeec2bd442b/env/env_posix.cc#L171). (Note that this will be combined with the `umask` so the files will generally not be world-readable; group-readability will depend on system configuration.)

---

*[pkg/storage/replica_sideload_disk.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohLK1yaytqrmt5J6ic:-KohLK1yaytqrmt5J6id:b-g597n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_disk.go#L59)):*
> ```Go
> 	}
> 	// File does not exist yet.
> 	return ioutil.WriteFile(filename, contents, 0600)
> ```

There's technically a bit of a race here between the Stat and WriteFile. We could avoid it by inlining WriteFile and adding `O_EXCL` to the open flags (and remove the `Stat` call, so there's only a few net lines of added code)

---

*[pkg/storage/replica_sideload_disk.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohLbO2BWo9w_WeuOue:-KohLbO2BWo9w_WeuOuf:bnritnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_disk.go#L66)):*
> ```Go
> 	if _, err := os.Stat(filename); err == nil {
> 		// File exists.
> 	} else if !os.IsNotExist(err) {
> ```

So if Stat returns a not-exist error, we continue to ReadFile anyway? That seems weird. Why even do the stat?

---

*[pkg/storage/replica_sideload_inmem.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohIyPKJUjoO35OvIxY:-KohIyPKJUjoO35OvIxZ:b-6ew9pc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_inmem.go#L30)):*
> ```Go
> }
> 
> type inMemSideloadStorage struct {
> ```

Any changes in the inmem version or just code movement?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-:-KohQ4-_e4uU37PVJd9m:bsxg0xj)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Added tests and brushed things up, PTAL! cc @danhhz

---

Review status: 0 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica.go, line 622 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohHPOOxabP3dZi-853:-Kos3bcFxYybnQsEOEP7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica.go#L622)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Nit: I'm not sure where this usage of "trivial" to mean "zero" came from, but it annoys me. 
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KoRMfajgq4-YZ7nCIve-r1-623:-Kos3_LBFliPx9PAgHB_:bxk3sns) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica.go#L623)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

How did we get a non-nil `r.raftMu.sideloaded` at the beginning of `init`?
</blockquote></details>

Moved this to correct location now.

---

*[pkg/storage/replica.go, line 695 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KoSmlK3P3UyGSnzqN8f:-Kos3geAE1e7k90p0xEn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica.go#L695)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes.
</blockquote></details>

Done.

---

*[pkg/storage/replica_sideload_disk.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohJdHaRalL7gDtTkSj:-Kos5AVv1TVeYpk6zYQ2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_disk.go#L41)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

For consistency with rocksdb, we should use permissions 0755 for directories and [0644 for files](https://github.com/facebook/rocksdb/blob/56656e12d67d8a63f1e4c4214da9feeec2bd442b/env/env_posix.cc#L171). (Note that this will be combined with the `umask` so the files will generally not be world-readable; group-readability will depend on system configuration.)
</blockquote></details>

Done.

---

*[pkg/storage/replica_sideload_disk.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohLK1yaytqrmt5J6ic:-Kos5BV1KALkVFj0h-4r:bux6j32) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_disk.go#L59)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There's technically a bit of a race here between the Stat and WriteFile. We could avoid it by inlining WriteFile and adding `O_EXCL` to the open flags (and remove the `Stat` call, so there's only a few net lines of added code)
</blockquote></details>

All operations are serialized through raftMu, so I won't go all-in right now. I'll keep it in mind though because I'm considering moving sideload storage truncation out of the Raft goroutine at some point.

---

*[pkg/storage/replica_sideload_disk.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohLbO2BWo9w_WeuOue:-Kos5rl9ySpqSF1nAn5J:b-ph8q72) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_disk.go#L66)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

So if Stat returns a not-exist error, we continue to ReadFile anyway? That seems weird. Why even do the stat?
</blockquote></details>

There was a bug here and also you're right -- much better to just call `ReadFile` directly.

---

*[pkg/storage/replica_sideload_inmem.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KohIyPKJUjoO35OvIxY:-Kos5vrMgmrbP3e4pope:b-th3856) ([raw file](https://github.com/cockroachdb/cockroach/blob/4264d65bf6f72bc3f1a5a7cea2a7fe9a2a1c6144/pkg/storage/replica_sideload_inmem.go#L30)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Any changes in the inmem version or just code movement?
</blockquote></details>

Just code movement initially and cosmetic changes in later diffs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-:-Kos6_KwmPh_W6AC5SVn:b-1miqmr)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-:-KoscJRSlmlpsmzSOFbc:bb6ywmx)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I didn't look at the tests closely

---

Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 811 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-Kosj7FqQ-dSsThickFV:-Kosj7FqQ-dSsThickFW:b-2idq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d0a03664a414124f2923f9c9bbdc1d46c552d1/pkg/storage/replica.go#L811)):*
> ```Go
> }
> 
> // setReplicaIDLocked requires that the replica lock is held.
> ```

and raft lock?

---

*[pkg/storage/replica_sideload_disk.go, line 49 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KosjI1bdbvMZUxZGrFa:-KosjI1bdbvMZUxZGrFb:b-nqyb4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d0a03664a414124f2923f9c9bbdc1d46c552d1/pkg/storage/replica_sideload_disk.go#L49)):*
> ```Go
> }
> 
> func (imss *diskSideloadStorage) createDir() error {
> ```

the receiver names of all these don't match the type name (copy pasta presumably). how about just using `s` to avoid this problem altogether?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-:-KoskhoM8CPjqLWfYBdO:b-3xxrcc)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: implement disk-based sideloaded storage

Tests are TODO. Out for early review since it's indispensable for serious
testing of RESTORE (the in-memory storage does not get ejected eagerly
enough since the Raft log queue is unaware of sideloaded data, so you
quickly get OOMs).

cc @danhhz

tschottdorf commented:
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/replica.go, line 811 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-Kosj7FqQ-dSsThickFV:-Kosm5uG6DvdVCVabcnX:b11x3hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d0a03664a414124f2923f9c9bbdc1d46c552d1/pkg/storage/replica.go#L811)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

and raft lock?
</blockquote></details>

Removed the comment.

---

*[pkg/storage/replica_sideload_disk.go, line 49 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-KosjI1bdbvMZUxZGrFa:-KosnAjFA2aMx4axgAHy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55d0a03664a414124f2923f9c9bbdc1d46c552d1/pkg/storage/replica_sideload_disk.go#L49)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

the receiver names of all these don't match the type name (copy pasta presumably). how about just using `s` to avoid this problem altogether?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTRs!

---

Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16924#-:-KosnBxmOpwYHWDTAzIk:bz5mair)*
<!-- Sent from Reviewable.io -->
